Framework: introduce sections middleware

samouri opened the pull request:
**Background**

Calypso is comprised of a lot of code.  Many megabytes of javascript.  In webapps, every byte of feature means slower initial load time for our end users because they need to download (or have cached) the entire app every time they access it. At least thats how things used to be until the concept of code-splitting came along.

code-splitting is a concept that says you can load part of the app intially, and then load other parts as you need it.  Conventially this was done via `require.ensure` calls and more recently the [import()](https://webpack.js.org/api/module-methods/#import) api.  Anywhere that webpack see either of those functions it will create what is called a split-point and spin off a new chunk that gets asynchronously loaded. 



Note: the parameter to `import()` must be a _literal_ string representing a module's path and cannot be an expression that is evaluated. This is because webpack needs to be able to statically determine exactly which module to split at compile time. For example:

``` javascript
// valid
import('lib/warn').then(warn => warn('this works!'))
const n = 'n';
import('./lib/war' + n ).then( warn => warn('this also works but bundles anything that matches ./lib/war**'));

// invalid
const moduleToLoad = 'lib/warn';
import(moduleToLoad).then( warn => warn('this does not work'))

```



**How Calypso uses code-splitting**

Some time long ago Calypso-ers invented our own homegrown concept built on top of this concept of code splitting called a section.  A section can be thought of as a sub-application within Calypso -- things like the Reader, Stats, Domains, or the Store. They are an extra layer of abstraction on top of direct calls to `import()` and add in routing with `page`.  Most of our section definitions live in [client/wordpress-com](https://github.com/Automattic/wp-calypso/blob/master/client/wordpress-com.js) and take this shape:

```javascript
// example section definition
{
	name: 'reader', // name of section
	paths: [ '/', '/read' ], // paths that page.js should listen for in order to transition to this section
	module: 'reader', // module that represents the entry of this section (reader/index.js)
	secondary: true, // should the sidebar be present
	group: 'reader', // one of (Reader | Sites) and determines the masterbar layout
}
```



These section definitions on their own do nothing.  It is up to a middleware to actually take these section definitions and make the `page` calls, invoke  `import()`, manage which sections have been loaded and which haven't, and also handle section loading errors.  Where does this middleware exist?  It kind of doesn't...it is generated on the fly by a custom webpack loader in [server/bundler/loader](https://github.com/Automattic/wp-calypso/blob/master/server/bundler/loader.js).  A webpack loader is a piece of code that takes in a file and spits back out a new file _at compile time_.  So what this one in particular does is take in the the section definitions, and for each one repeats the same bit of code (by copying a string template over and over) to generate a piece of code that looks like this for each section (psuedocode): 

```js
// repeat this code once per section.
page( section.path, () => {
  if ( alreadyHaveSectionCached ) {
    switchToSection( sectionCache.get( section.path ) );
    return;
  }
  import( section.module )
    .then( module => {
    	sectionCache.put( section.path, module );	
    	switchToSection( module );
  	} )
    .catch( () => showLoadingError() );
})
```

what about if code-splitting was turned off in our configs? Well the loader would output something a little bit different:

```javascript
// repeat once per section.
page( section.path, () => {
  swithToSection( require( section.module ) );
})
```



**The problem**

While the solution has taken us pretty far, there are three issues i'd like to highlight with it:

1. Bundle size: The actual code generated is about 135Kb. This is incidental size and not necessary.  There is no reason to repeat the same logic over and over seeing as loops are pretty good at that. This PR shows that it can be gotten down to ~1Kb.
2. Build Speed: The loader as written takes much longer than I expected during the webpack client build. On my Mac, `master` takes about 34 seconds to build whereas with the loader removed  it only takes 24 seconds.
3. Maintainability: using a heavy webpack loader like this one introduces a lot of complexity into our build process that I believe is purely incidental.  On top of that, the current loader has no tests and we'd be hard-pressed to write tests for a bunch of concatenated strings.  It doesn't make sense to write js inside of interpolated strings as opposed to just writing js.

**Proposal**

Move 99.99% of this logic into a fairly standard [calypso middleware](https://github.com/Automattic/wp-calypso/blob/master/client/boot/common.js#L194). After all of my investigating, there is only one bit that _must be_ inside of a webpack loader as opposed to plain js.  And that is the `import` vs. `require` statement.  We need to keep this as written by the loader so that we have the freedom to enable/disable code-splitting easily.  That said, we can totally minimize what the loader is responsible for and make it just do a very simple string transformation. If we change the section definition to look like this: 

```javascript
// example section definition
{
  ...,
  module: 'reader',
  load: LOADER( 'reader' ),
  ...
}
```

Then the loader can be as simple as:

```javascript
const split = config.isEnabled('code-splitting')
export default content => content.replace( 'LOADER', split ? 'import' : 'require' );
```

---

Note the implementation is still a WIP.

~~client build times: 34s --> 24~~
~~section.js size: 135K -->  1.3K~~

todo:
- [x] fix chunk names so they are the same as before this pr
- [x] address todos in code
- [x] make work with codesplitting turned off
- [x] tests
- [x] handle multiple modules in the same section name.

Thank you @yurynix for all your work in simplifying the webpack loader.

yurynix commented on client/sections-middleware.js:
```diff
@@ -0,0 +1,179 @@
+}
+
+const moduleToImport = {
+	'jetpack-connect': import( /* webpackChunkName: "'jetpack-connect'" */ 'jetpack-connect' ),
```
What happens here is once the browser hit that, it'll fetch all sections, that's now how the loader used to work

yurynix commented:
ICFY says this branch increases build size:
http://iscalypsofastyet.com/branch?branch=try/sections/remove-loader
```
Delta:
chunk                                                         stat_size             parsed_size               gzip_size
build                                                         +448529 B    (+9.0%)    +242413 B    (+9.9%)     +93915 B   
(+19.2%)
```

It also changes current chunks structure...

samouri commented:
@yurynix: its still a work-in-progress!
I have a number issues to still iron out

dmsnell commented on client/sections-middleware.js:
```diff
@@ -0,0 +1,179 @@
+
+function maybeLoadCSS( sectionName ) {
+	//eslint-disable-line no-unused-vars
+	const section = find( sectionsWithCss, elem => elem.name === sectionName );
```
while you're at it, you might like `lodash.matchesProperty()`

```js
const section = find( sectionsWithCss, matchesProperty( 'name', sectionName ) )
```

samouri replied:
i'm pretty sure i could even do:
```js
const section = find( secttionWithCss, { name: sectionName } )
```

dmsnell commented on client/sections-middleware.js:
```diff
@@ -0,0 +1,179 @@
+			section.paths.forEach( path => createPageDefinition( path, section ) )
+		);
+	},
+};
```
how about named exports? seems like they'd make sense here

```js
export const get = () => sections

export const load = () => sections.forEach(
    section => section.paths.forEach( path => createPathDefinition( path, section ) )
)
```

samouri replied:
will do :)

yurynix commented on server/bundler/sections-loader.js:
```diff
@@ -0,0 +1,43 @@
+function addModuleImportToSections( { sections, shouldSplit } ) {
+	sections.forEach( section => {
+		const loaderFunction = ( () => require( /* webpackChunkName: 'CHUNK_NAME' */ section.module ) )
+			.toString()
```
❤️

yurynix commented:
Love it, great work @samouri , thank you!

yurynix commented on client/sections-middleware.js:
```diff
@@ -0,0 +1,112 @@
+	if ( path === '/' ) {
+		return path;
+	}
+	return new RegExp( '^' + path + '(/.*)?$' );
```
Lets [escape](https://github.com/Automattic/wp-calypso/pull/21255/files#diff-b2907500f646f51484f42b858d1d63b6R8) `path` with `escapeRegExp`

ockham replied:
Isn't `escapeRegExp` redundant, given https://github.com/Automattic/wp-calypso/pull/21349/files#diff-afb3c8d4236d194988c31a97e465126dR108 - 110?

On a sidenote, IIRC we're using the RegEx here so that a section definition like https://github.com/Automattic/wp-calypso/blob/abb552dfba238d93315629bebb4939bcaaaf5c6a/client/wordpress-com.js#L205 matches both `/theme` and `/theme/*`. However, a RegEx seems almost overkill to do this, when `page.js` also comes with its ownpseudo-regular syntax. Maybe we should just map each string in a section definition's `paths` array into two -- one unchanged, and one with `/*` appended.

yurynix commented on server/bundler/sections-loader.js:
```diff
@@ -0,0 +1,64 @@
+	const sections = require( this.resourcePath );
+	const sectionsWithCss = withCss( sections );
+
+	console.error(
```
Guess you forgot that =)

ockham commented:
/cc @ehg

ockham commented:
BTW great writeup @samouri! We should consider using your PR desc as the base for a doc on our bundler loader (unless it's already covered by https://github.com/Automattic/wp-calypso/blob/master/server/bundler/README.md, haven't read that in a while).

samouri commented:
Good news and bad news.

Bad news is that my initial numbers were off due to a faulty implementation.

- webpack build speed improvements are nothing to write home about.  it was only so much faster because it was missing so many modules.
- the `sections.js` module is still getting much smaller 🎉 .  I'll have more exact numbers when i put this into "needs review" 

The good news is that this PR is very close to ready!

samouri commented:
I'm putting this as needs review now. Heres the summary of what's changed between this and master:
1. 99.9% of the logic that used to exist as a series of concatenated strings in the loader has been moved to `sections-middleware`. The overall logic hasn't been heavily modified.
2. the loader still exists but its responsibility has been minimized. All it does now is flesh out the section definitions with two more properties: `load` and `css`. 

ICFY: having trouble getting results....will edit this comment with bundle size diffs soon

Places for more improvement:
- the file `sections-middleware` could still use some simplification as well as refactoring so that it can be unit tested.

samouri commented:
This PR also helps with the esmodules effort since it converts a `require` into an `import` in `sections.js`.  Otherwise we'd need to update the require to look like `require().default` before disabling cjs modules.  

cc @jsnajdr @dmsnell

yurynix commented:
This is good to go IMO, great work ! :+1:

samouri commented:
@jsnajdr pointed out that I may have broken SSR..and this probably has because i renamed a couple functions

samouri commented:
I think the only place the isomorphic section loader is used is `server/pages/index.js` and that remains unchanged...so this might actually be good. 

@ockham: any chance you can sanity check my logic?

---
To Test:
1. run the app in local dev and make sure clicking around works
2. run the app without code splitting and make sure it works as expected: `DISABLE_FEATURES=code-splitting npm start`

If both of those work i'd feel comfortable merging and reverting if it breaks in wpcalypso -- this change shouldn't be stage dependent.

samouri commented:
I was having trouble getting SSR to work so I figure I'd clean it up a bit while I was there making changes anyway.

I've gone ahead and removed the loader the server had in favor of reusing the exact same one as the client. There were two points that informed how I decided to implement 
the server:
1. should use `require` instead of a dynamic import (no chunk splitting).
2. should only bundle along isomorphic sections

Because of that i've added two flags to the loader: `forceRequire` and `onlyIsomorphic`.

ockham commented:
❤️ this. It's awesome to see the dedicated server-side loader removed. I had a look at the SSR related bits, and they make sense to me; I also verified that the Theme Showcase is still properly SSR'd.

And ooh, there's even a unit test for the loader 😍

ockham commented on server/bundler/sections-loader.js:
```diff
@@ -0,0 +1,69 @@
+const getOptions = loaderUtils.getOptions;
+
+/*
+ * This sections-loader has two one responsibilties: adding import statements and css details.
```
`two one`? 😄

samouri replied:
whoopsies. i'll cross out _one_

ockham commented on server/bundler/sections-loader.js:
```diff
@@ -0,0 +1,69 @@
+ * It also searches for sections with a `css` key and adds in css filename corresponding to that css chunk id.
+ * Technically this could happen as part of sections-middleware, but it relies upon nodejs crypto that would pull
+ * in huge modules to the client, so for now its being done as part of the loader.
+ */
```
Great explanation 👍 Awesome that it even describes what it would take to remove the loader.

ockham commented on client/sections-middleware.js:
```diff
@@ -0,0 +1,103 @@
+function preload( sectionName ) {
+	maybeLoadCSS( sectionName );
+	const filteredSections = filter( sections, { name: sectionName } );
+	if ( isEmpty( sections ) ) {
```
I might be mistaken, but don't we want to check for emptiness of `filteredSections` here?

samouri replied:
nice catch!

this hasn't come up during testing because this should never really catch...why would we ever try to preload a section that doesn't exist.

We can refactor logic another time though, i'll fix that now

ockham commented on client/sections-middleware.js:
```diff
@@ -0,0 +1,103 @@
+preloadHub.on( 'preload', preload );
+
+function activateSection( sectionDefinition, context, next ) {
+	const dispatch = context.store.dispatch;
```
Nitpick -- Can we destructure here? Or is that part of the language spec not available at this stage?

samouri replied:
🤷‍♂️

sirreal replied:
If the question is "Does Caylpso's Node version (8.9.3 at the moment) support destructuring assignment?" the answer is "Yes."

ockham commented on client/sections-middleware.js:
```diff
@@ -0,0 +1,103 @@
+
+	page( pathRegex, function( context, next ) {
+		const envId = sectionDefinition.envId;
+		const dispatch = context.store.dispatch;
```
https://github.com/Automattic/wp-calypso/pull/21349/files#r162147980
